File: backend/api/curation/applier.py
"""Pure journal applier.

Takes a JournalEntry and a GraphBuildState, returns a new state with the
op applied. No I/O. Used by the repository layer (transactional persist)
and by the orchestrator's preview endpoint (in-memory dry run).
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass, field
from typing import Any

from api.domain.curation import JournalEntry, JournalOp
from api.domain.graph import Edge, EdgeType, Node
from api.domain.types import Id, new_id
from api.strategies.state import GraphBuildState

from .ops import (
    AddEdgePayload,
    DeleteEdgePayload,
    DeleteNodePayload,
    EditEdgePayload,
    MergeNodesPayload,
    MoveToCommunityPayload,
    RetypeNodePayload,
    SetSummaryPayload,
    SplitNodePayload,
    UpdateNodeNamePayload,
    parse_payload,
)


class JournalApplyError(RuntimeError):
    """Apply could not proceed because the entry references an object
    no longer in the state (e.g. edit on a deleted edge). Caller decides
    whether to skip or fail the run.
    """


@dataclass(frozen=True, slots=True)
class AffectedSet:
    """What downstream pipelines need to refresh after this op.

    Used by the vector-outbox writer (re-embed `node_ids`), the summary
    refresher agent (Phase 3 — re-summarize `community_ids`), and the
    F7 layered viewer (selectively invalidate cached layouts).
    """

    node_ids: frozenset[Id] = field(default_factory=frozenset)
    edge_ids: frozenset[Id] = field(default_factory=frozenset)
    community_ids: frozenset[Id] = field(default_factory=frozenset)

    def union(self, other: "AffectedSet") -> "AffectedSet":
        return AffectedSet(
            node_ids=self.node_ids | other.node_ids,
            edge_ids=self.edge_ids | other.edge_ids,
            community_ids=self.community_ids | other.community_ids,
        )


# ---- public API ----


def apply_journal_op(
    state: GraphBuildState, entry: JournalEntry
) -> GraphBuildState:
    """Return a new state with `entry` applied. The entry itself is
    appended to state.journal. Pure: caller's state object is unchanged.
    """

    payload = parse_payload(entry.op, entry.payload)
    nodes = list(state.nodes)
    edges = list(state.edges)

    match entry.op:
        case JournalOp.MERGE_NODES:
            assert isinstance(payload, MergeNodesPayload)
            nodes, edges = _apply_merge_nodes(nodes, edges, payload)
        case JournalOp.SPLIT_NODE:
            assert isinstance(payload, SplitNodePayload)
            nodes, edges = _apply_split_node(nodes, edges, payload, entry)
        case JournalOp.RETYPE_NODE:
            assert isinstance(payload, RetypeNodePayload)
            nodes = _update_node(nodes, payload.node_id, type=payload.new_type)
        case JournalOp.MOVE_TO_COMMUNITY:
            assert isinstance(payload, MoveToCommunityPayload)
            edges = _apply_move_to_community(edges, payload, entry)
        case JournalOp.EDIT_EDGE:
            assert isinstance(payload, EditEdgePayload)
            edges = _apply_edit_edge(edges, payload)
        case JournalOp.DELETE_EDGE:
            assert isinstance(payload, DeleteEdgePayload)
            edges = [e for e in edges if e.id != payload.edge_id]
        case JournalOp.DELETE_NODE:
            assert isinstance(payload, DeleteNodePayload)
            nodes = [n for n in nodes if n.id != payload.node_id]
            edges = [
                e
                for e in edges
                if e.source_node_id != payload.node_id
                and e.target_node_id != payload.node_id
            ]
        case JournalOp.ADD_EDGE:
            assert isinstance(payload, AddEdgePayload)
            edges.append(Edge.model_validate(payload.edge))
        case JournalOp.SET_SUMMARY:
            assert isinstance(payload, SetSummaryPayload)
            nodes = _update_node(nodes, payload.node_id, summary=payload.summary)
        case JournalOp.UPDATE_NODE_NAME:
            assert isinstance(payload, UpdateNodeNamePayload)
            nodes = _update_node(nodes, payload.node_id, name=payload.name)

    return GraphBuildState(nodes=nodes, edges=edges, journal=state.journal + [entry])


def replay_journal(
    base_state: GraphBuildState, entries: Iterable[JournalEntry]
) -> GraphBuildState:
    """Apply entries in order. Used to materialize a current state from
    a parent variant + journal stream (Phase 2.4 undo and Phase 2.x
    cross-variant recovery scenarios).
    """

    state = base_state
    for entry in entries:
        state = apply_journal_op(state, entry)
    return state


def affected_set(
    state_before: GraphBuildState, entry: JournalEntry
) -> AffectedSet:
    """Compute the downstream impact of `entry` against the pre-apply
    state. Returns the set of nodes that need re-embedding, edges that
    were touched, and communities whose membership shifted.
    """

    payload = parse_payload(entry.op, entry.payload)
    node_index = state_before.node_index()

    match entry.op:
        case JournalOp.MERGE_NODES:
            assert isinstance(payload, MergeNodesPayload)
            return _affected_merge(payload, state_before, node_index)
        case JournalOp.SPLIT_NODE:
            assert isinstance(payload, SplitNodePayload)
            community = _community_of(payload.original_id, state_before)
            new_node_ids = _new_split_ids(payload, entry)
            return AffectedSet(
                node_ids=frozenset({payload.original_id, *new_node_ids}),
                community_ids=frozenset({community} if community else ()),
            )
        case JournalOp.RETYPE_NODE:
            assert isinstance(payload, RetypeNodePayload)
            community = _community_of(payload.node_id, state_before)
            return AffectedSet(
                node_ids=frozenset({payload.node_id}),
                community_ids=frozenset({community} if community else ()),
            )
        case JournalOp.MOVE_TO_COMMUNITY:
            assert isinstance(payload, MoveToCommunityPayload)
            communities = {payload.to_community_id}
            if payload.from_community_id is not None:
                communities.add(payload.from_community_id)
            else:
                # Reconstruct the from-community from current state.
                inferred = _community_of(payload.node_id, state_before)
                if inferred:
                    communities.add(inferred)
            return AffectedSet(
                node_ids=frozenset({payload.node_id}),
                community_ids=frozenset(communities),
            )
        case JournalOp.EDIT_EDGE:
            assert isinstance(payload, EditEdgePayload)
            edge = next((e for e in state_before.edges if e.id == payload.edge_id), None)
            if edge is None:
                return AffectedSet(edge_ids=frozenset({payload.edge_id}))
            communities = _communities_of_endpoints(edge, node_index, state_before)
            return AffectedSet(
                edge_ids=frozenset({edge.id}),
                node_ids=frozenset({edge.source_node_id, edge.target_node_id}),
                community_ids=frozenset(communities),
            )
        case JournalOp.DELETE_EDGE:
            assert isinstance(payload, DeleteEdgePayload)
            edge = next((e for e in state_before.edges if e.id == payload.edge_id), None)
            if edge is None:
                return AffectedSet(edge_ids=frozenset({payload.edge_id}))
            communities = _communities_of_endpoints(edge, node_index, state_before)
            return AffectedSet(
                edge_ids=frozenset({edge.id}),
                node_ids=frozenset({edge.source_node_id, edge.target_node_id}),
                community_ids=frozenset(communities),
            )
        case JournalOp.DELETE_NODE:
            assert isinstance(payload, DeleteNodePayload)
            community = _community_of(payload.node_id, state_before)
            touched_edges = frozenset(
                e.id
                for e in state_before.edges
                if e.source_node_id == payload.node_id
                or e.target_node_id == payload.node_id
            )
            return AffectedSet(
                node_ids=frozenset({payload.node_id}),
                edge_ids=touched_edges,
                community_ids=frozenset({community} if community else ()),
            )
        case JournalOp.ADD_EDGE:
            assert isinstance(payload, AddEdgePayload)
            spec = payload.edge
            src = Id(spec["source_node_id"]) if not isinstance(spec.get("source_node_id"), Id) else spec["source_node_id"]
            tgt = Id(spec["target_node_id"]) if not isinstance(spec.get("target_node_id"), Id) else spec["target_node_id"]
            communities = set()
            for nid in (src, tgt):
                community = _community_of(nid, state_before)
                if community:
                    communities.add(community)
            return AffectedSet(
                node_ids=frozenset({src, tgt}),
                community_ids=frozenset(communities),
            )
        case JournalOp.SET_SUMMARY | JournalOp.UPDATE_NODE_NAME:
            node_id = Id(payload.node_id)  # type: ignore[union-attr]
            return AffectedSet(node_ids=frozenset({node_id}))


# ---- internals ----


def _apply_merge_nodes(
    nodes: list[Node],
    edges: list[Edge],
    payload: MergeNodesPayload,
) -> tuple[list[Node], list[Edge]]:
    absorbed = set(payload.absorbed_ids)
    if payload.survivor_id in absorbed:
        raise JournalApplyError("survivor cannot be in absorbed list")

    nodes = [n for n in nodes if n.id not in absorbed]
    redirect = {aid: payload.survivor_id for aid in absorbed}
    return nodes, _redirect_edges(edges, redirect)


def _apply_split_node(
    nodes: list[Node],
    edges: list[Edge],
    payload: SplitNodePayload,
    entry: JournalEntry,
) -> tuple[list[Node], list[Edge]]:
    nodes = [n for n in nodes if n.id != payload.original_id]
    new_node_ids = _new_split_ids(payload, entry)
    new_nodes = []
    for spec, node_id in zip(payload.new_nodes, new_node_ids, strict=True):
        spec_with_id = dict(spec)
        spec_with_id.setdefault("id", node_id)
        spec_with_id.setdefault("graph_variant_id", entry.graph_variant_id)
        new_nodes.append(Node.model_validate(spec_with_id))
    nodes.extend(new_nodes)

    if not new_node_ids:
        # No new nodes — edges that touched the original are dropped.
        return nodes, [
            e
            for e in edges
            if e.source_node_id != payload.original_id
            and e.target_node_id != payload.original_id
        ]

    fallback = new_node_ids[0]
    redirect = {payload.original_id: fallback}
    rerouted: list[Edge] = []
    for e in edges:
        target_for_edge = payload.edge_redirect.get(str(e.id), redirect.get(e.source_node_id))
        if e.source_node_id == payload.original_id or e.target_node_id == payload.original_id:
            new_src = (
                target_for_edge
                if e.source_node_id == payload.original_id
                else e.source_node_id
            )
            new_tgt = (
                target_for_edge
                if e.target_node_id == payload.original_id
                else e.target_node_id
            )
            if new_src is None or new_tgt is None or new_src == new_tgt:
                continue
            rerouted.append(
                e.model_copy(update={"source_node_id": new_src, "target_node_id": new_tgt})
            )
        else:
            rerouted.append(e)
    return nodes, rerouted


def _apply_move_to_community(
    edges: list[Edge],
    payload: MoveToCommunityPayload,
    entry: JournalEntry,
) -> list[Edge]:
    out = [
        e
        for e in edges
        if not (
            e.type == EdgeType.MEMBER_OF
            and e.source_node_id == payload.node_id
            and (
                payload.from_community_id is None
                or e.target_node_id == payload.from_community_id
            )
        )
    ]
    out.append(
        Edge(
            graph_variant_id=entry.graph_variant_id,
            type=EdgeType.MEMBER_OF,
            source_node_id=payload.node_id,
            target_node_id=payload.to_community_id,
        )
    )
    return out


def _apply_edit_edge(edges: list[Edge], payload: EditEdgePayload) -> list[Edge]:
    target = next((e for e in edges if e.id == payload.edge_id), None)
    if target is None:
        raise JournalApplyError(f"edit_edge: {payload.edge_id} not found")
    bad = set(payload.updates) - set(Edge.model_fields)
    if bad:
        raise JournalApplyError(f"edit_edge: unknown fields {sorted(bad)!r}")
    return [
        e if e.id != payload.edge_id else e.model_copy(update=payload.updates)
        for e in edges
    ]


def _update_node(nodes: list[Node], node_id: Id, **updates: Any) -> list[Node]:
    found = False
    out = []
    for n in nodes:
        if n.id == node_id:
            found = True
            out.append(n.model_copy(update=updates))
        else:
            out.append(n)
    if not found:
        raise JournalApplyError(f"node {node_id} not found")
    return out


def _redirect_edges(edges: list[Edge], redirect: dict[Id, Id]) -> list[Edge]:
    out: list[Edge] = []
    seen: set[tuple[Id, Id, str]] = set()
    for e in edges:
        src = redirect.get(e.source_node_id, e.source_node_id)
        tgt = redirect.get(e.target_node_id, e.target_node_id)
        if src == tgt:
            continue
        key = (src, tgt, e.type.value)
        if key in seen:
            continue
        seen.add(key)
        if src == e.source_node_id and tgt == e.target_node_id:
            out.append(e)
        else:
            out.append(
                e.model_copy(update={"source_node_id": src, "target_node_id": tgt})
            )
    return out


def _community_of(node_id: Id, state: GraphBuildState) -> Id | None:
    for e in state.edges:
        if e.type == EdgeType.MEMBER_OF and e.source_node_id == node_id:
            return e.target_node_id
    return None


def _communities_of_endpoints(
    edge: Edge,
    node_index: dict[Id, Node],
    state: GraphBuildState,
) -> set[Id]:
    out: set[Id] = set()
    for nid in (edge.source_node_id, edge.target_node_id):
        community = _community_of(nid, state)
        if community:
            out.add(community)
    return out


def _new_split_ids(payload: SplitNodePayload, entry: JournalEntry) -> list[Id]:
    """Stable IDs for the pieces produced by a split. Uses ids embedded
    in the payload spec when present, otherwise mints fresh ones — but
    deterministically per (entry id, position) so replay yields the
    same nodes.
    """

    out: list[Id] = []
    for i, spec in enumerate(payload.new_nodes):
        if "id" in spec and spec["id"] is not None:
            out.append(Id(spec["id"]) if not isinstance(spec["id"], type(new_id())) else spec["id"])
        else:
            out.append(new_id())
    return out


def _affected_merge(
    payload: MergeNodesPayload,
    state: GraphBuildState,
    node_index: dict[Id, Node],
) -> AffectedSet:
    touched_ids = {payload.survivor_id, *payload.absorbed_ids}
    communities: set[Id] = set()
    for nid in touched_ids:
        community = _community_of(nid, state)
        if community:
            communities.add(community)
    return AffectedSet(
        node_ids=frozenset(touched_ids),
        community_ids=frozenset(communities),
    )

File: packages/graph-core/src/engine/applier.ts
import type { Node, Edge, Verification } from "../domain/graph.js";
import { granularityOf, FACTMETA_KEY, type Layer } from "../domain/graph.js";
import type { JournalEntry } from "../domain/journal.js";
import { newId, type Id } from "../domain/types.js";
import { communityOf, type GraphState } from "./state.js";
import { parsePayload, type PayloadOf } from "./payloads.js";

/** Apply could not proceed because the entry references an object no
 *  longer in the state (e.g. edit on a deleted edge). */
export class JournalApplyError extends Error {}

export interface AffectedSet {
  nodeIds: Set<Id>;
  edgeIds: Set<Id>;
  communityIds: Set<Id>;
}

const emptyAffected = (): AffectedSet => ({
  nodeIds: new Set(),
  edgeIds: new Set(),
  communityIds: new Set(),
});

// ── public API ──

/** Return a NEW state with `entry` applied and appended to the journal.
 *  Pure: the caller's state is untouched. */
export function applyJournalOp(
  state: GraphState,
  entry: JournalEntry,
): GraphState {
  let nodes = [...state.nodes];
  let edges = [...state.edges];

  switch (entry.op) {
    case "merge_nodes": {
      const p = parsePayload("merge_nodes", entry.payload);
      [nodes, edges] = applyMerge(nodes, edges, p);
      break;
    }
    case "split_node": {
      const p = parsePayload("split_node", entry.payload);
      [nodes, edges] = applySplit(nodes, edges, p, entry);
      break;
    }
    case "retype_node": {
      const p = parsePayload("retype_node", entry.payload);
      nodes = updateNode(nodes, p.nodeId, { type: p.newType });
      break;
    }
    case "move_to_community": {
      const p = parsePayload("move_to_community", entry.payload);
      edges = applyMove(edges, p, entry);
      break;
    }
    case "edit_edge": {
      const p = parsePayload("edit_edge", entry.payload);
      edges = applyEditEdge(edges, p);
      break;
    }
    case "delete_edge": {
      const p = parsePayload("delete_edge", entry.payload);
      edges = applyDeleteEdge(edges, p, entry);
      break;
    }
    case "delete_node": {
      const p = parsePayload("delete_node", entry.payload);
      nodes = nodes.filter((n) => n.id !== p.nodeId);
      edges = edges.filter(
        (e) => e.sourceId !== p.nodeId && e.targetId !== p.nodeId,
      );
      break;
    }
    case "add_edge": {
      const p = parsePayload("add_edge", entry.payload);
      edges = [...edges, coerceEdge(p.edge, entry.graphId)];
      break;
    }
    case "set_summary": {
      const p = parsePayload("set_summary", entry.payload);
      nodes = updateNode(nodes, p.nodeId, { summary: p.summary });
      break;
    }
    case "update_node_name": {
      const p = parsePayload("update_node_name", entry.payload);
      nodes = updateNode(nodes, p.nodeId, { name: p.name });
      break;
    }
    case "set_layer": {
      const p = parsePayload("set_layer", entry.payload);
      nodes = updateNode(nodes, p.nodeId, {
        layer: p.newLayer,
        granularity: p.granularity ?? granularityOf(p.newLayer),
      });
      break;
    }
    case "set_verified": {
      const p = parsePayload("set_verified", entry.payload);
      const v: Verification | null = p.verified
        ? { by: entry.actor, at: entry.createdAt, asOf: p.asOf ?? null, axis: p.axis ?? "tx" }
        : null;
      if (p.edgeId) edges = updateEdge(edges, p.edgeId, { verified: v });
      else if (p.nodeId && p.attrKey) nodes = updateFactMeta(nodes, p.nodeId, p.attrKey, { verified: v });
      else if (p.nodeId) nodes = updateNode(nodes, p.nodeId, { verified: v });
      break;
    }
    case "set_confidence": {
      const p = parsePayload("set_confidence", entry.payload);
      if (p.edgeId) edges = updateEdge(edges, p.edgeId, { confidenceScore: p.confidence });
      else if (p.nodeId && p.attrKey)
        nodes = updateFactMeta(nodes, p.nodeId, p.attrKey, { confidence: p.confidence });
      break;
    }
    case "set_attribute": {
      const p = parsePayload("set_attribute", entry.payload);
      nodes = nodes.map((n) => {
        if (n.id !== p.nodeId) return n;
        const attrs = { ...n.attributes };
        if (p.value === null || p.value === undefined) delete attrs[p.key];
        else attrs[p.key] = p.value;
        return { ...n, attributes: attrs };
      });
      break;
    }
  }

  return { nodes, edges, journal: [...state.journal, entry] };
}

/** Apply entries in order. Used to materialize current state from a base
 *  state + its journal stream (and to power revert = replay-without-last). */
export function replayJournal(
  base: GraphState,
  entries: Iterable<JournalEntry>,
): GraphState {
  let state = base;
  for (const entry of entries) state = applyJournalOp(state, entry);
  return state;
}

/** Downstream impact of `entry` against the pre-apply state: which nodes
 *  need re-embedding, which edges were touched, which communities shifted. */
export function affectedSet(
  before: GraphState,
  entry: JournalEntry,
): AffectedSet {
  const out = emptyAffected();
  const addCommunity = (nid: Id) => {
    const c = communityOf(before, nid);
    if (c) out.communityIds.add(c);
  };

  switch (entry.op) {
    case "merge_nodes": {
      const p = parsePayload("merge_nodes", entry.payload);
      for (const nid of [p.survivorId, ...p.absorbedIds]) {
        out.nodeIds.add(nid);
        addCommunity(nid);
      }
      break;
    }
    case "split_node": {
      const p = parsePayload("split_node", entry.payload);
      out.nodeIds.add(p.originalId);
      for (const nid of newSplitIds(p, entry)) out.nodeIds.add(nid);
      addCommunity(p.originalId);
      break;
    }
    case "retype_node": {
      const p = parsePayload("retype_node", entry.payload);
      out.nodeIds.add(p.nodeId);
      addCommunity(p.nodeId);
      break;
    }
    case "move_to_community": {
      const p = parsePayload("move_to_community", entry.payload);
      out.nodeIds.add(p.nodeId);
      out.communityIds.add(p.toCommunityId);
      if (p.fromCommunityId) out.communityIds.add(p.fromCommunityId);
      else addCommunity(p.nodeId);
      break;
    }
    case "edit_edge":
    case "delete_edge": {
      const edgeId =
        entry.op === "edit_edge"
          ? parsePayload("edit_edge", entry.payload).edgeId
          : parsePayload("delete_edge", entry.payload).edgeId;
      out.edgeIds.add(edgeId);
      const edge = before.edges.find((e) => e.id === edgeId);
      if (edge) {
        out.nodeIds.add(edge.sourceId);
        out.nodeIds.add(edge.targetId);
        addCommunity(edge.sourceId);
        addCommunity(edge.targetId);
      }
      break;
    }
    case "delete_node": {
      const p = parsePayload("delete_node", entry.payload);
      out.nodeIds.add(p.nodeId);
      for (const e of before.edges) {
        if (e.sourceId === p.nodeId || e.targetId === p.nodeId)
          out.edgeIds.add(e.id);
      }
      addCommunity(p.nodeId);
      break;
    }
    case "add_edge": {
      const p = parsePayload("add_edge", entry.payload);
      const src = String(p.edge.sourceId ?? "");
      const tgt = String(p.edge.targetId ?? "");
      for (const nid of [src, tgt]) {
        if (nid) {
          out.nodeIds.add(nid);
          addCommunity(nid);
        }
      }
      break;
    }
    case "set_summary":
    case "update_node_name":
    case "set_layer":
    case "set_verified":
    case "set_confidence":
    case "set_attribute": {
      const pl = entry.payload as Record<string, unknown>;
      const nid = String(pl.nodeId ?? "");
      const eid = String(pl.edgeId ?? "");
      if (nid) {
        out.nodeIds.add(nid);
        addCommunity(nid);
      }
      if (eid) {
        out.edgeIds.add(eid);
        const edge = before.edges.find((e) => e.id === eid);
        if (edge) {
          out.nodeIds.add(edge.sourceId);
          out.nodeIds.add(edge.targetId);
        }
      }
      break;
    }
  }
  return out;
}

// ── internals ──

const EDGE_FIELDS = new Set<string>([
  "type",
  "sourceId",
  "targetId",
  "weight",
  "relation",
  "explanation",
  "confidence",
  "provenance",
  "attributes",
  "validFrom",
  "validTo",
  "txFrom",
  "txTo",
  "invalidation",
]);

function applyMerge(
  nodes: Node[],
  edges: Edge[],
  p: PayloadOf<"merge_nodes">,
): [Node[], Edge[]] {
  const absorbed = new Set(p.absorbedIds);
  if (absorbed.has(p.survivorId))
    throw new JournalApplyError("survivor cannot be in absorbed list");

  // Preserve every alias so nothing is lost: the absorbed nodes' names + their
  // existing aliases (and, when the survivor is renamed, its old name) all roll
  // up onto the survivor's `attributes.aliases`.
  const aliasOf = (n: Node): string[] => {
    const a = (n.attributes as Record<string, unknown>)?.aliases;
    return Array.isArray(a) ? a.map((x) => String(x)) : [];
  };
  const aliasSet = new Set<string>();
  const addAlias = (s: string | null | undefined) => {
    const v = (s ?? "").trim();
    if (v) aliasSet.add(v);
  };
  for (const n of nodes) {
    if (!absorbed.has(n.id)) continue;
    addAlias(n.name);
    for (const a of aliasOf(n)) addAlias(a);
  }

  let seen = false;
  nodes = nodes.map((n) => {
    if (n.id !== p.survivorId) return n;
    seen = true;
    for (const a of aliasOf(n)) addAlias(a); // keep the survivor's own aliases
    const finalName = p.newName != null ? (p.newName as string) : n.name;
    if (p.newName != null) addAlias(n.name); // renamed → old name is an alias
    aliasSet.delete(finalName); // a node is never its own alias
    const aliases = [...aliasSet];
    const attributes =
      aliases.length > 0
        ? { ...n.attributes, aliases }
        : n.attributes;
    return { ...n, name: finalName, attributes };
  });
  if (!seen)
    throw new JournalApplyError(
      `merge_nodes: survivor ${p.survivorId} not found`,
    );

  nodes = nodes.filter((n) => !absorbed.has(n.id));
  const redirect = new Map<Id, Id>();
  for (const aid of absorbed) redirect.set(aid, p.survivorId);
  return [nodes, redirectEdges(edges, redirect)];
}

function applySplit(
  nodes: Node[],
  edges: Edge[],
  p: PayloadOf<"split_node">,
  entry: JournalEntry,
): [Node[], Edge[]] {
  nodes = nodes.filter((n) => n.id !== p.originalId);
  const newIds = newSplitIds(p, entry);
  const created = p.newNodes.map((spec, i) =>
    coerceNode({ id: newIds[i], ...spec }, entry.graphId),
  );
  nodes = [...nodes, ...created];

  if (newIds.length === 0) {
    return [
      nodes,
      edges.filter(
        (e) => e.sourceId !== p.originalId && e.targetId !== p.originalId,
      ),
    ];
  }

  const fallback = newIds[0]!;
  const out: Edge[] = [];
  for (const e of edges) {
    const touches = e.sourceId === p.originalId || e.targetId === p.originalId;
    if (!touches) {
      out.push(e);
      continue;
    }
    const target = p.edgeRedirect[e.id] ?? fallback;
    const newSrc = e.sourceId === p.originalId ? target : e.sourceId;
    const newTgt = e.targetId === p.originalId ? target : e.targetId;
    if (!newSrc || !newTgt || newSrc === newTgt) continue;
    out.push({ ...e, sourceId: newSrc, targetId: newTgt });
  }
  return [nodes, out];
}

function applyMove(
  edges: Edge[],
  p: PayloadOf<"move_to_community">,
  entry: JournalEntry,
): Edge[] {
  const out = edges.filter(
    (e) =>
      !(
        e.type === "member_of" &&
        e.sourceId === p.nodeId &&
        (p.fromCommunityId == null || e.targetId === p.fromCommunityId)
      ),
  );
  out.push({
    id: newId(),
    graphId: entry.graphId,
    type: "member_of",
    sourceId: p.nodeId,
    targetId: p.toCommunityId,
    confidence: "extracted",
    provenance: [],
    attributes: {},
  });
  return out;
}

function applyEditEdge(edges: Edge[], p: PayloadOf<"edit_edge">): Edge[] {
  const target = edges.find((e) => e.id === p.edgeId);
  if (!target) throw new JournalApplyError(`edit_edge: ${p.edgeId} not found`);
  const bad = Object.keys(p.updates).filter((k) => !EDGE_FIELDS.has(k));
  if (bad.length)
    throw new JournalApplyError(`edit_edge: unknown fields ${bad.join(", ")}`);
  return edges.map((e) =>
    e.id === p.edgeId ? ({ ...e, ...p.updates } as Edge) : e,
  );
}

function applyDeleteEdge(
  edges: Edge[],
  p: PayloadOf<"delete_edge">,
  entry: JournalEntry,
): Edge[] {
  if (p.reason == null) return edges.filter((e) => e.id !== p.edgeId);
  const at = p.supersededAt ?? entry.createdAt;
  const invalidation = {
    ingestionEventId: p.ingestionEventId ?? null,
    at,
    reason: p.reason,
    supersededByEdgeId: null,
    auto: entry.actor.startsWith("agent:"),
  };
  return edges.map((e) =>
    e.id === p.edgeId ? { ...e, txTo: at, invalidation } : e,
  );
}

function updateNode(nodes: Node[], id: Id, updates: Partial<Node>): Node[] {
  let found = false;
  const out = nodes.map((n) => {
    if (n.id === id) {
      found = true;
      return { ...n, ...updates };
    }
    return n;
  });
  if (!found) throw new JournalApplyError(`node ${id} not found`);
  return out;
}

function updateEdge(edges: Edge[], id: Id, updates: Partial<Edge>): Edge[] {
  let found = false;
  const out = edges.map((e) => {
    if (e.id === id) {
      found = true;
      return { ...e, ...updates };
    }
    return e;
  });
  if (!found) throw new JournalApplyError(`edge ${id} not found`);
  return out;
}

/** Set per-attribute fact metadata (confidence/verified) under the node's
 *  reserved __factmeta__ attribute. */
function updateFactMeta(
  nodes: Node[],
  id: Id,
  attrKey: string,
  patch: { confidence?: number | null; verified?: Verification | null },
): Node[] {
  return updateNode(nodes, id, {} as Partial<Node>).map((n) => {
    if (n.id !== id) return n;
    const meta = { ...((n.attributes[FACTMETA_KEY] as Record<string, unknown>) ?? {}) };
    meta[attrKey] = { ...((meta[attrKey] as Record<string, unknown>) ?? {}), ...patch };
    return { ...n, attributes: { ...n.attributes, [FACTMETA_KEY]: meta } };
  });
}

/** Redirect edge endpoints; drop self-loops; dedup identical
 *  (src,tgt,type,relation), keeping the first. */
function redirectEdges(edges: Edge[], redirect: Map<Id, Id>): Edge[] {
  const out: Edge[] = [];
  const seen = new Set<string>();
  for (const e of edges) {
    const src = redirect.get(e.sourceId) ?? e.sourceId;
    const tgt = redirect.get(e.targetId) ?? e.targetId;
    if (src === tgt) continue;
    const key = `${src}|${tgt}|${e.type}|${e.relation ?? ""}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(
      src === e.sourceId && tgt === e.targetId
        ? e
        : { ...e, sourceId: src, targetId: tgt },
    );
  }
  return out;
}

/** Stable ids for split pieces: use ids embedded in the spec when present,
 *  else mint fresh ones. (Caller passes deterministic ids via the journal
 *  payload so replay yields the same nodes.) */
function newSplitIds(p: PayloadOf<"split_node">, _entry: JournalEntry): Id[] {
  return p.newNodes.map((spec) => {
    const id = (spec as Record<string, unknown>).id;
    return typeof id === "string" && id ? id : newId();
  });
}

/** Build a full Node from a partial spec, filling sane defaults. */
export function coerceNode(spec: Record<string, unknown>, graphId: Id): Node {
  const layer = (spec.layer as Layer) ?? "entity";
  return {
    id: (spec.id as string) ?? newId(),
    graphId: (spec.graphId as string) ?? graphId,
    layer,
    type: (spec.type as string) ?? "MISC",
    granularity: (spec.granularity as number) ?? granularityOf(layer),
    name: (spec.name as string) ?? "",
    summary: (spec.summary as string | null) ?? null,
    attributes: (spec.attributes as Record<string, unknown>) ?? {},
    provenance: (spec.provenance as Node["provenance"]) ?? [],
    x: (spec.x as number | null) ?? null,
    y: (spec.y as number | null) ?? null,
    pinned: (spec.pinned as boolean) ?? false,
    validFrom: (spec.validFrom as string | null) ?? null,
    validTo: (spec.validTo as string | null) ?? null,
    txFrom: (spec.txFrom as string | null) ?? null,
    txTo: (spec.txTo as string | null) ?? null,
  };
}

/** Build a full Edge from a partial spec, filling sane defaults. */
export function coerceEdge(spec: Record<string, unknown>, graphId: Id): Edge {
  return {
    id: (spec.id as string) ?? newId(),
    graphId: (spec.graphId as string) ?? graphId,
    type: (spec.type as Edge["type"]) ?? "entity_relation",
    sourceId: spec.sourceId as string,
    targetId: spec.targetId as string,
    weight: (spec.weight as number | null) ?? null,
    relation: (spec.relation as string | null) ?? null,
    explanation: (spec.explanation as string | null) ?? null,
    confidence: (spec.confidence as Edge["confidence"]) ?? "extracted",
    provenance: (spec.provenance as Edge["provenance"]) ?? [],
    attributes: (spec.attributes as Record<string, unknown>) ?? {},
    validFrom: (spec.validFrom as string | null) ?? null,
    validTo: (spec.validTo as string | null) ?? null,
    txFrom: (spec.txFrom as string | null) ?? null,
    txTo: (spec.txTo as string | null) ?? null,
    invalidation: (spec.invalidation as Edge["invalidation"]) ?? null,
  };
}
